wgengine/tstun/tun_test.go

```diff
@@ -0,0 +1,354 @@
+// Copyright (c) 2020 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+package tstun
+
+import (
+	"bytes"
+	"testing"
+
+	"github.com/tailscale/wireguard-go/tun/tuntest"
+	"tailscale.com/types/logger"
+	"tailscale.com/wgengine/filter"
+	"tailscale.com/wgengine/packet"
+)
+
+func udp(src, dst packet.IP, sport, dport uint16) []byte {
+	header := packet.UDPHeader{
+		IPHeader: packet.IPHeader{
+			SrcIP: src,
+			DstIP: dst,
+			IPID:  0,
+		},
+		SrcPort: sport,
+		DstPort: dport,
+	}
+	return packet.Generate(&header, []byte("udp_payload"))
```

Comment:
generally unless it's an output parameter or otherwise being acted upon (like `fmt.Fprintf(&buf, ...`), don't pass the address of a local variable to a func and instead keep it as a pointer everywhere in the func. (so `header := &packet.UDPHeader{`). Otherwise it sometimes leads to people making value copies of things that shouldn't be and having diverged copies of variables in places.

The compiler can/should still allocate things even when you write `t := &T{....}` and pass `t` to a func where it doesn't escape.

(doesn't really matter usually, of course, and this is just a test. just pointing out a minor style thing)